### PR TITLE
Wayland: Restore tablet support and handle multiple tools

### LINKED ISSUE #88333: [Wayland] Godot doesn't register inputs from graphic tablet on GNOME
### Tested versions

Reproducible only in v4.3.dev3.official [36e943b6b], since it's first version that introduced Wayland support.

### System information

Fedora Linux 39.20240214.0 (Silverblue), Linux 6.7.4-200.fc39.x86_64, AMD Radeon RX 570 Series, GNOME 45.4 on Wayland session, tablet in use Wacom One Pen Display 13, Forward+ renderer

### Issue description

Godot editor with enforced Wayland `run/platforms/linuxbsd/prefer_wayland = true` completely ignores inputs from Wacom tablet on Wayland session, including clicking and moving with pen. Same happens with running projects with `display/display_server/driver.linuxbsd = wayland`. With this options turned off, everything works as expected.

### Steps to reproduce

1. Change `run/platforms/linuxbsd/prefer_wayland` in editor setting to `true`.
2. Restart Godot.
3. Now Godot won't accept any inputs from tablet via pen.

### Minimal reproduction project (MRP)

Same could be achieved with running projects.
1. Create empty project.
2. Change project setting `display/display_server/driver.linuxbsd` to `wayland`.
3. Add node and attach script:
```gdscript
extends Node

func _input(event):
	print(event.as_text())
```
It generate properly generates inputs with mouse:
```
Mouse motion at position ((702, 633)) with velocity ((-0.000042, 0.000042))
Mouse motion at position ((697, 635)) with velocity ((-0.000625, 0.00025))
```
But moving around and clicking with pen generates nothing.
[wayland_input_wacom.zip](https://github.com/godotengine/godot/files/14284484/wayland_input_wacom.zip)

### PR DESCRIPTION
**Pull request by deralmas**
Fixes #88333.

This code was already partially there, although heavily incomplete and nowadays commented out.

It got broken after the `WaylandThread` refactor and I didn't bother to bring it over, preferring to `#if 0` it into oblivion for the time being as I don't have a tablet/pen which support an eraser and tilt reporting.

This commit brings it back and adds proper multi-tool support (needed for eraser detection) thanks to @winston-yallow, who could test this code with their more capable tablet.

---

I also noticed some occasional weird segfaults even without even having a tablet installed which I haven't been able to isolate/replicate reliably, so I'm not able to tell if they're a new thing or not.

Although it has been confirmed that eraser and tilt are properly reported now, considering the above, some further testing would be wise.

Here's a nice lil' visualizer I quickly made for this PR, please give it a spin if you have a drawing tablet: https://github.com/riteo/godot-tablet-visualizer

(I wonder if we could make this a demo or something, it's real nifty :D )

### CODE REVIEW

**Comment by Calinou** [comment]
> (I wonder if we could make this a demo or something, it's real nifty :D )

I think it makes sense to submit to https://github.com/godotengine/godot-demo-projects, so it can be used as a troubleshooting tool like we do with the Joypads demo.

**Review by akien-mga — APPROVED** [review]
Code looks good to me overall. Haven't tested yet (I have an older laptop with a touchscreen and pen, I'll see if it gets recognized as a tablet). *Edit:* Pen is out of battery and I don't have any AAAA batteries laying around, so I can't test now :P

I didn't check in detail but the `memnew`/`memdelete` stuff around `TabletToolState` probably needs some caution to avoid double-free or memory leaks (it's good to remember that if it needs to be freed after use, any `ERR_FAIL` kind of early return need to ensure it gets freed first too).

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
> Should be good to go already I think, as if there's no godot state set (e.g. libdecor or whatever messing around with our display) it should return `null`, noopping.

Is `memdelete(nullptr)` safe? I see we have an unused `memdelete_notnull` macro. We often do `if (thing) memdelete(thing);` if there's a possibility for the `thing` to be null.

**Comment by akien-mga** [comment]
> Edit: wait, what about `memdelete_not_null`?

Since it was never used in the codebase, I think for now we should just do the `if (thing) memdelete(thing);` and discuss with the core team what to do more generally with this pattern.

We could refactor the whole codebase to use `memdelete_not_null`, or drop the macro and keep doing things ad hoc, or evaluate whether it would be that bad to just add a null check in `memdelete` itself.

File: platform/linuxbsd/wayland/wayland_thread.cpp [hunk 1]
```diff
@@ -2160,179 +2152,190 @@ void WaylandThread::_wp_primary_selection_source_on_cancelled(void *data, struct
...
-	while (it) {
-		struct zwp_tablet_tool_v2 *tool = it->get();
+	List<struct zwp_tablet_tool_v2 *>::Element *E = ss->tablet_tools.find(zwp_tablet_tool_v2);
 
-		if (tool == zwp_tablet_tool_v2) {
-			zwp_tablet_tool_v2_destroy(tool);
-			ss->tablet_tools.erase(it);
-			break;
+	if (E && E->get()) {
+		TabletToolState *state = wp_tablet_tool_get_state(tool);
```
Comment (akien-mga) [inline platform/linuxbsd/wayland/wayland_thread.cpp:2204]: There seems to be an issue after refactoring this, `tool` doesn't exist here.

Comment (akien-mga) [reply platform/linuxbsd/wayland/wayland_thread.cpp:2204]: Looks like you dropped `struct zwp_tablet_tool_v2 *tool = E->get();`.

**Comment by akien-mga** [comment]
I built and tested af273a1efde392b4ca479b285df5d3047d9504a0 + the hotfix to bring back `tool` (which you then committed as https://github.com/godotengine/godot/compare/af273a1efde392b4ca479b285df5d3047d9504a0..b01a36b3cd04499d959bbd169dec719e088a420a), so I pretty much tested the latest version of this PR already, ahead of its commit date :P

Also out of battery (AAAA, who uses those anymore aside from Ubisoft...) for my touchscreen pen so I couldn't actually check that the feature works :)

**Review by winston-yallow — APPROVED** [review]
I've tested the most recent version (b01a36b3cd04499d959bbd169dec719e088a420a), everything seems to work perfectly.

### Device
Wacom Intuos M Pro (2017)

### Tilted backwards / right, medium pressure, normal tip:
image

### Tilted backwards / slightly left, high pressure, eraser tip:
image

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**